Create Dask resource

kinghuang opened the pull request:
This PR creates a Dask resource, which manages a Dask client and optional cluster. The resource is to be used by solids and the Dask DataFrame type's materializer to compute Dask graphs.

The resource can be configured to connect to a pre-existing cluster by its scheduler address, or create a cluster on demand.

Here are example resource configs for typical use cases.

**Connect to an existing Dask cluster via its scheduler:**

```yaml
resources:
  dask:
    config:
      client:
        name: my-dagster-pipeline
        address: tcp://dask-scheduler-here:8786
```

**Create a local cluster with 4 workers, and 1 thread per worker:**

The resource will create a Cluster object, and pass it as the address option to the client.

```yaml
resources:
  dask:
    config:
      client:
        name: my-dagster-pipeline
      cluster:
        local:
          n_workers: 4
          threads_per_worker: 1
```

The DataFrame type *does not* require a Dask resource. But, if a resource is provided under the `dask` key, the type's materializer (if configured) will run with the resource's client as the current client. Otherwise, the global client and scheduler will apply.

Related to the above, the `compute` field is removed from all the materialization options. Providing a way to not compute a Dask DataFrame materialization does not make sense, as the future is never returned in the pipeline and cannot be computed at a later time. Materializations must be computed if specified.

Solids using the Dask resource for computation should consider making the resource client the current client by using the [as_current()](https://distributed.dask.org/en/latest/api.html#distributed.Client.as_current) resource manager.

```python
@solid(…)
def some_solid(context):
  with context.resources.dask.client.as_current():
    …
```

This PR carries the commits from #2821, and is meant to be merged after it.

alangenfeld commented:
can you add some basic tests to make sure this doesn't get broken over time?

alangenfeld commented:
other than that I think this is fine, we just need to settle out the schema changes on the PR this depends on

kinghuang commented:
Added a couple of tests for local clusters created via the resource.

I've also changed the implementation to *not require* a Dask resource on the DataFrame type so that it is possible to simply run with Dask's global config (as before). And, I'm thinking of making `set_as_default` an option instead of forcing it to be `False`.

kinghuang commented:
@alangenfeld This is ready for review. Thanks!

kinghuang commented:
Found an error. Putting this back on draft for now.

kinghuang commented:
Updated. The code formatting issues identified by Black in `dagster_dask/data_frame.py` have been fixed in #2888.

alangenfeld commented on python_modules/libraries/dagster-dask/dagster_dask/data_frame.py:
```diff
@@ -288,8 +287,7 @@
         "options": {
             "path": (Any, True, "Path to a target filename."),
             "key": (String, True, "Datapath within the files."),
-            "compute": (Bool, False, "Whether or not to execute immediately."),
```
did you mean to remove this? just concerned about breaking changes

kinghuang replied:
Yes, all the `compute` options have been removed, because they don't make sense in the context of materialization configs. Setting `compute=False` will cause Dask to return a future, which in the materialization flow will never be computed. This would result in `AssetMaterialization` objects being yielded for assets that don't actually exist.

alangenfeld replied:
Can you add a note to `Changes.md` at the repo root - something like
```
## 0.9.6 (Upcoming)
**Breaking Changes**
* [dagster-dask] removed the `compute` options key which would result in un-executed futures if used 
```

but add some more context on where that key is

kinghuang replied:
Added a note to the changelog.

kinghuang replied:
Rebased and updated for post-0.9.6.

alangenfeld commented on python_modules/libraries/dagster-dask/dagster_dask/resources.py:
```diff
@@ -0,0 +1,143 @@
+                raise ValueError(f"Unknown cluster type “{cluster_type}”.")
+
+            # Import the cluster module by name, and get the cluster Class.
+            cluster_module = import_module(cluster_meta["module"])
```
a bit concerned about the late import here, and how missing dependency failures will manifest - but I can't come up with a solution that really solves or that. This same problem exists in the executor as well.

kinghuang replied:
Few ideas:

1. Each cluster type's field description could mention their required module.
2. Is there a hook for validating resource configs? That might be a good time to check if a module exists.
3. The `dagster_dask` module could declare `extra_requires` for the modules. Though, I'm not sure if this is something this module wants to take on, considering Dask's `distributed` module doesn't already do it, and it might not be obvious to users.

kinghuang replied:
As an aside, I'm aware that the resource and executor will have separate implementations for configuring and creating Dask clients and clusters with this PR. I'm thinking of following up with another PR later to bring them together under a common implementation.

kinghuang replied:
I've added the module name to each cluster config's description. For example, the `kube` type's description now says:

> Kubernetes cluster config. Requires dask_kubernetes.

alangenfeld commented:
Great, those further tweaks all sound great. Will get this merged once I get a clean buildkite run.

An aside - not sure if you caught this yet, but you can run `make black` and `make isort` at the repo root to catch formatting issues. Also `make pylint` but that is very very slow, so maybe just reference the cli args and point at the files you changed.

kinghuang commented:
Ah, I've been running those manually, and obviously getting it wrong compared to the automated checks. I'll give that a try!

kinghuang commented:
@alangenfeld Do I need some sort of config files to run `black` and `isort`? I just rebased and ran `make black` on the project root and it reformatted 348 files across the project. Similar for `make isort`.

kinghuang commented:
I've discarded all the changes outside of `dagster_dask` and pushed an update in the meantime.

alangenfeld commented:
ah shoot, ya you can see what versions of `black` and `isort` we are set to here https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dev-requirements.txt should have mentioned that

kinghuang commented:
Ah, much better. Unfortunately, it's undone some of the changes the other run of black applied. Updated! 😅
